frappe/desk/form/assign_to.py

```diff
@@ -40,56 +41,58 @@ def add(args=None):
-		AND `status`='Open'
-		AND `owner`=%(assign_to)s""", args):
-		frappe.throw(_("Already in user's To Do list"), DuplicateToDoError)
-	else:
-		from frappe.utils import nowdate
-
-		if not args.get('description'):
-			args['description'] = _('Assignment for {0} {1}').format(args['doctype'], args['name'])
-
-		d = frappe.get_doc({
-			"doctype":"ToDo",
-			"owner": args['assign_to'],
+	for assign_to in json.loads(args.get("assign_to")):
+		filters = {
 			"reference_type": args['doctype'],
 			"reference_name": args['name'],
-			"description": args.get('description'),
-			"priority": args.get("priority", "Medium"),
 			"status": "Open",
-			"date": args.get('date', nowdate()),
-			"assigned_by": args.get('assigned_by', frappe.session.user),
-			'assignment_rule': args.get('assignment_rule')
-		}).insert(ignore_permissions=True)
-
-		# set assigned_to if field exists
-		if frappe.get_meta(args['doctype']).get_field("assigned_to"):
-			frappe.db.set_value(args['doctype'], args['name'], "assigned_to", args['assign_to'])
-
-		doc = frappe.get_doc(args['doctype'], args['name'])
-
-		# if assignee does not have permissions, share
-		if not frappe.has_permission(doc=doc, user=args['assign_to']):
-			frappe.share.add(doc.doctype, doc.name, args['assign_to'])
-			frappe.msgprint(_('Shared with user {0} with read access').format(args['assign_to']), alert=True)
-
-		# make this document followed by assigned user
-		follow_document(args['doctype'], args['name'], args['assign_to'])
+			"owner": assign_to
+		}
 
-	# notify
-	notify_assignment(d.assigned_by, d.owner, d.reference_type, d.reference_name, action='ASSIGN',\
-			description=args.get("description"))
+		if frappe.get_all("ToDo", filters=filters):
+			if not args.get("bulk_assign"):
+				frappe.throw(_("Already in user's ToDo list"), DuplicateToDoError)
+		else:
+			from frappe.utils import nowdate
+
+			if not args.get('description'):
+				args['description'] = _('Assignment for {0} {1}').format(args['doctype'], args['name'])
+
+			d = frappe.get_doc({
+				"doctype": "ToDo",
+				"owner": assign_to,
+				"reference_type": args['doctype'],
+				"reference_name": args['name'],
+				"description": args.get('description'),
+				"priority": args.get("priority", "Medium"),
+				"status": "Open",
+				"date": args.get('date', nowdate()),
+				"assigned_by": args.get('assigned_by', frappe.session.user),
+				'assignment_rule': args.get('assignment_rule')
+			}).insert(ignore_permissions=True)
+
+			# set assigned_to if field exists
+			if frappe.get_meta(args['doctype']).get_field("assigned_to"):
+				frappe.db.set_value(args['doctype'], args['name'], "assigned_to", assign_to)
+
+			doc = frappe.get_doc(args['doctype'], args['name'])
+
+			# if assignee does not have permissions, share
+			if not frappe.has_permission(doc=doc, user=assign_to):
+				frappe.share.add(doc.doctype, doc.name, assign_to)
+				frappe.msgprint(_('Shared with user {0} with read access').format(assign_to, alert=True))
```

Comment:
this will keep stacking since it will be running in a loop? there should be a better way of showing this too right?

Reply:
Every assignee has to be notified. The assignes wont be too many to keep stacking afaik

Reply:
makes sense, but the point i am trying to make in here is when it gets stacked, it'll show the message "Shared with user \<email\> with read access" on every line, instead a message which goes something like:

```
Shared with the following users with read access:

mail@example.com
mail2@example.com
mailN@example.com
```

I think the latter one looks neater and better to look at, no?